### PR TITLE
fix(fetch HK sample sheet for validation new demux post processing flow)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

This PR changes the location of the sample sheet used in validation during the new demultuplexing post processing flow.
With these changes the sample sheet location is in housekeeper and not the `flow_cells/flow_cell_dir_name`. 

These changes go paired with that upon sample sheet creation for a given flow cell, the sample sheet is directly stored in housekeeper.

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/demultiplex/validation.py [hunk 1]
```diff
@@ -83,8 +83,10 @@ def validate_flow_cell_delivery_status(flow_cell_output_directory: Path, force:
 
 
 def is_flow_cell_ready_for_postprocessing(
```
Comment (islean) [inline cg/meta/demultiplex/validation.py:85]: Feels a bit weird to me that this is a void function rather than a boolean given the naming. Also makes the test assertions a bit weird - `assert is_flow_cell_ready_for_postprocessing is None` feels closer to `assert not is_flow_cell_ready_for_postprocessing` than the opposite.

File: tests/meta/demultiplex/test_validation.py [hunk 2]
```diff
@@ -172,22 +173,29 @@ def test_is_flow_cell_ready_for_delivery_false(tmp_path: Path):
...
+    sample_sheet_path = Path(
+        bcl2fastq_flow_cell_dir, DemultiplexingDirsAndFiles.SAMPLE_SHEET_FILE_NAME
+    )
+    sample_sheet_path.touch()
+    flow_cell._sample_sheet_path_hk = sample_sheet_path
 
     # WHEN validating the existence of the sample sheet
     # THEN it should not raise an error
-    assert validate_sample_sheet_exists(tmp_path) is None
+    assert validate_sample_sheet_exists(flow_cell=flow_cell) is None
```
Comment (islean) [inline tests/meta/demultiplex/test_validation.py:198]: Sort of similar to above although "validate_sample_sheet_exists" is perhaps a bit better. Still feels like "assert is None" should be a failure though, and I would probably prefer a boolean return type on validate_sample_sheet_exists.

Comment (islean) [reply tests/meta/demultiplex/test_validation.py:198]: A bit unsure though. I could be convinced that validate functions should be void and only raise an exception if anything goes wrong.

**Review by islean — APPROVED** [review]
Looks nice. Just a bit confused by some return types, but that should not be a part of this PR either way.

File: tests/meta/demultiplex/test_validation.py (hunk 2 above)
Comment (ChrOertlin) [reply tests/meta/demultiplex/test_validation.py:198]: I agree, it was a bit confusing when I looked at it as well. @diitaz93 and I were talking about a demux post processing refactoring when we established the new flow. We will have a look at it then.

**Comment by ChrOertlin** [comment]
> 

I agree, it was a bit confusing when I looked at it as well. @diitaz93 and I were talking about a demux post processing refactoring when we established the new flow. We will have a look at it then.

File: tests/meta/demultiplex/test_validation.py [hunk 3]
```diff
@@ -172,22 +173,29 @@ def test_is_flow_cell_ready_for_delivery_false(tmp_path: Path):
...
+    flow_cell._sample_sheet_path_hk = None
     # WHEN validating the existence of the sample sheet
     # THEN it should raise a FlowCellError
     with pytest.raises(FlowCellError):
-        validate_sample_sheet_exists(tmp_path)
+        validate_sample_sheet_exists(flow_cell=flow_cell)
 
 
-def test_validate_sample_sheet_exists_no_error(tmp_path: Path):
+def test_validate_sample_sheet_exists_no_error(bcl2fastq_flow_cell_dir: Path):
```
Comment (diitaz93) [inline tests/meta/demultiplex/test_validation.py:186]: I think testing the correct functionality of the test should be just test_function, but feel free to disregard
```suggestion
def test_validate_sample_sheet_exists(bcl2fastq_flow_cell_dir: Path):
```

**Review by diitaz93 — APPROVED** [review]
Nice 👍

File: cg/meta/demultiplex/validation.py (hunk 1 above)
Comment (diitaz93) [reply cg/meta/demultiplex/validation.py:85]: agree

File: tests/meta/demultiplex/test_validation.py (hunk 2 above)
Comment (diitaz93) [reply tests/meta/demultiplex/test_validation.py:198]: This could be discussed in coding guidelines meetings: return type of validators. I think for this one is ok to return None. The one above should return bool (`is_flow_cell_ready_for_postprocessing`)

File: cg/meta/demultiplex/validation.py [hunk 4]
```diff
@@ -57,14 +60,11 @@ def is_flow_cell_ready_for_delivery(flow_cell_directory: Path) -> bool:
...
-        flow_cell_run_directory, DemultiplexingDirsAndFiles.SAMPLE_SHEET_FILE_NAME
-    )
+def validate_sample_sheet_exists(flow_cell: FlowCellDirectoryData) -> None:
+    sample_sheet_path: Path = flow_cell.get_sample_sheet_path_hk()
     if not sample_sheet_path.exists():
-        raise FlowCellError(
-            f"Sample sheet {sample_sheet_path} does not exist in flow cell run directory."
-        )
+        raise FlowCellError(f"Sample sheet {sample_sheet_path} does not exist in housekeeper.")
+    LOG.debug(f"Found sample sheet {sample_sheet_path} in housekeeper.")
```
Comment (Vince-janv) [inline cg/meta/demultiplex/validation.py:67]: I think just `if not sample_sheet_path:` should be enough. This risks running into `None type object has not attribute exists()`.

**Review by Vince-janv — APPROVED** [review]
Nice, I like having this in the validate command 👍

File: cg/meta/demultiplex/validation.py (hunk 4 above)
Comment (islean) [reply cg/meta/demultiplex/validation.py:67]: I agree that we should have the null check but I do not think this is sufficient - would this approach not miss if the file on Hasta was missing although the db entry is filled in? 

I would prefer `if (not sample_sheet_path or not sample_sheet_path.exists())` and use the lazy loading of "or" in Python.

File: cg/meta/demultiplex/validation.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/meta/demultiplex/validation.py:85]: I agree with you both. Let's keep that for another refactoring. I think it will require a bit too much change right now.

File: tests/meta/demultiplex/test_validation.py (hunk 2 above)
Comment (ChrOertlin) [reply tests/meta/demultiplex/test_validation.py:198]: In line with the comment above. I also disagree with the return type of the function. I will leave it for now though.

File: cg/meta/demultiplex/validation.py (hunk 4 above)
Comment (ChrOertlin) [reply cg/meta/demultiplex/validation.py]: fixed to check for both.

**Comment by ChrOertlin** [comment]
# CG demultiplex finish temporary looks for sample sheet in HK
image

# cg demultiplex finish temporary-all looks for sample sheet in hk
image

**Comment by ChrOertlin** [comment]
stage:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs/230824_A00187_0994_BHGW2VDRX3/Unaligned/Project_887803/Sample_ACC12589A10
done.
Log deploy... done.
cg, version 44.0.4

prod:
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs/230824_A00187_0994_BHGW2VDRX3/Unaligned/Project_887803/Sample_ACC12589A10
done.
Log deploy... done.
cg, version 44.0.4
